File: Exams/Resoluções de Exames by Carlos Martins/generic-synch/GenericSynchronizerMonitorStyle.cs
/**
 *  ISEL, LEIC, Concurrent Programming
 *
 *  Pseudo-code for implementing a generic synchronizer using the
 *  "Lampson & Redell monitor style" and .NET.
 *
 *  Realizations for the case of semaphore and message queue synchronizers.
 *
 *  Carlos Martins, October 2018
 */

#define EXCLUDE_PSEUDO_CODE

using System;
using System.Threading;
using System.Collections.Generic;

#if (!EXCLUDE_PSEUDO_CODE)

/**
  * Generic classes used for arguments and results
  */
  
public class SynchState {}
public class InitializeArgs {}
public class AcquireArgs {}
public class AcquireResult {}
public class ReleaseArgs {}

#endif

/**
 * This class implements a generic synchronizer using pseudo-code to express
 * data access and control synchronization (without considering the availability
 * of monitors).
 */

#if (!EXCLUDE_PSEUDO_CODE)

public class GenericSynchronizerMonitorStylePseudoCode {
	// lock that synchronizes the access to the mutable shared state
	private Lock _lock = new Lock();

    // queue where block threads are (also protected by "lock") needed to
    // implement control synchronization inherent to acquire/release semantics.
    private WaitQueue waitQueue = new WaitQueue();
	
	// synchronization state
	private SynchState synchState;
	
	// initialize the synchronizer
	public GenericSynchronizerMonitorStyle(InitializeState initialState) {
		initialize "synchState" according to information specified by "initialState";
	}
	
	// test if synchronization state allows an immediate acquire
	private bool CanAcquire(AcquireArgs acquireArgs) {
		returns true if "syncState" satisfies an immediate acquire according to "acquireArgs";
	}
	
	// executes the processing associated with a successful acquire 
	private AcquireResult AcquireSideEffect(AcquireArgs acquireArgs) {
		update "synchState" according to "acquireArgs" after a successful acquire;
		returns "the-proper-acquire-result"
	}
	
	// update synchronization state due to a release operation
	private void UpdateStateOnRelease(ReleaseArgs releaseArgs) {
		update "syncState" according to "releaseArgs";
	}
	
	public AcquireResult Acquire(AcquireArgs acquireArgs) {
		_lock.Acquire();
		try {
			while (!CanAcquire(acquireArgs)) {
				enqueue current thread on the "waitQueue" sensible to posterior wakeups;
				int depth = _lock.ReleaseAll();
				block current thread until it is waked up by a releaser thread;
				_lock.ReAcquire(depth);
			}
			return AcquireSideEffect(acquireArgs);
		} finally {
			_lock.Release();
		}
	}
	
	public void Release(ReleaseArgs releaseArgs) {
		_lock.Acquire();
		try {
			UpdateStateOnRelease(releaseArgs);
			wakeup all the blocked threads that can have its acquire satisfied with the new value of "syncState";
		} finally {
			_lock.Release();
		}
	}
}

#endif

/**
 * Generic synchronizer pseudo-code based on an *implicit .NET monitor*, with
 * support for timeout on the acquire operation.
 */

 #if (!EXCLUDE_PSEUDO_CODE)

public class GenericSynchronizerMonitorStyleImplicitMonitor {
    // implicit .NET monitor that suports the synchronzation of shared data access
    // and supports also the control synchronization.
    private readonly object monitor = new Object();
    
	// synchronization state
	private SynchState synchState;

    // initialize the synchronizer
	public GenericSynchronizerMonitorStyleImplicitMonitor(InitializeArgs initialArgs) {
        initialize "synchState" according to information specified by "initialArgs";
    }

    // returns true if synchronization state allows an immediate acquire
    private bool CanAcquire(AcquireArgs acquireArgs) {
        returns true if "syncState" satisfies an immediate acquire according to "acquireArgs";
    }

    // executes the processing associated with a successful acquire 
    private AcquireResult AcquireSideEffect(AcquireArgs acquireArgs) {
        update "synchState" according to "acquireArgs" after a successful acquire;
        returns "the-proper-acquire-result";
    }

    // update synchronization state due to a release operation
    private void UpdateStateOnRelease(ReleaseArgs releaseArgs) {
        update "syncState" according to "releaseArgs";
    }

    public bool Acquire(AcquireArgs acquireArgs, out AcquireResult result, int timeout = Timeout.Infinite) {
        lock(monitor) {
			if (CanAcquire(acquireArgs)) {
				result = AcquireSideEffect(acquireArgs);
				return true;
			}
			TimeoutHolder th = new TimeoutHolder(timeout);
			do {
				if ((timeout = th.Value) == 0) {
					result = default(AcquireResult);
					return false;
				}
				try {
                    Monitor.Wait(monitor, timeout);
				} catch (ThreadInterruptedException) {
                    // if notification was made with Monitor.Pulse, the single notification
					// may be lost if the blocking of the notified thread was interrupted,
					// so, if an acquire is possible, we notify another blocked thread, if any.
					if (CanAcquire(acquireArgs))
						Monitor.Pulse(monitor);
					throw;
				}
			} while (!CanAcquire(acquireArgs));
        	result = AcquireSideEffect(acquireArgs);
			return true;
		}
    }

    public void Release(ReleaseArgs releaseArgs) {
		lock(monitor) {
        	UpdateStateOnRelease(releaseArgs);
			Monitor.PulseAll(monitor);	/* or Monitor.Pulse if only one thread can have success in its acquire */
		}
    }
}

#endif

/**
 * Semaphore following the monitor style, using an *implicit .NET monitor*, with
 * support for timeout on the acquire operation.
 */
public class SemaphoreMonitorStyleImplicitMonitor {
    // implicit .NET monitor that suports the synchronzation of shared data access
    // and supports also the control synchronization.
    private readonly Object monitor = new Object();
	
	// synchronization state
	private int permits;
	
	// initialize the semaphore
	public SemaphoreMonitorStyleImplicitMonitor(int initial) {
		if (initial > 0)
			permits = initial;
	}
	
	// if the pending permits e equal or greater to the request,
	// we can acquire immediately
	private bool CanAcquire(int acquires) { return permits >= acquires; }
	
	// deduce the acquired permits
	private void AcquireSideEffect(int acquires) { permits -= acquires; }
	
	// take into account the released permits
	private void UpdateStateOnRelease(int releases) { permits += releases; }
	
	public bool Acquire(int acquires, int timeout) {
		lock(monitor) {
			if (CanAcquire(acquires)) {
                AcquireSideEffect(acquires);
				return true;
			}
			TimeoutHolder th = new TimeoutHolder(timeout);
			do {
				if ((timeout = th.Value) == 0)
					return false;
				Monitor.Wait(monitor, timeout);
			} while (!CanAcquire(acquires));
			AcquireSideEffect(acquires);
			return true;
		}
	}

	public void Release(int releases) {
		lock(monitor) {
			UpdateStateOnRelease(releases);
			Monitor.PulseAll(monitor);	/* the release can satisfy multiple acquires */
		}
	}
}

/**
 * Message queue following the monitor style, using a *implicit .NET monitor*,
 * with timeout support on the receive acquire operation.
 */
public class MessageQueueMonitorStyleImplicitMonitor<T> {
    // implicit .NET monitor that suports the synchronzation of shared data access
    // and supports also the control synchronization.
    private readonly Object monitor = new Object();
	
	// list of pending messages
 	private readonly LinkedList<T> pendingMessages = new LinkedList<T>();

    // initializes the message queue
	public MessageQueueMonitorStyleImplicitMonitor() {}

    // if there is an pending message, receive cam succeed immediately
    private bool CanReceive() { return pendingMessages.Count > 0; }

    // when a message is received, it must be removed from the pending message list
	private T ReceiveSideEffect() {
		T receivedMessage = pendingMessages.First.Value;
		pendingMessages.RemoveFirst();
		return receivedMessage;
    }

    // add the sent message to the pending messages list
    private void UpdateStateOnSend(T sentMessage) { pendingMessages.AddLast(sentMessage); }

    // receive the nest message
	public bool Receive(out T receivedMessage, int timeout = Timeout.Infinite) {
        lock(monitor) {
			if (CanReceive()) {
				receivedMessage = ReceiveSideEffect();
				return true;
			}
			TimeoutHolder th = new TimeoutHolder(timeout);
			do {
				if ((timeout = th.Value) == 0) {
					receivedMessage = default(T);
					return false;
				}
				try {
					Monitor.Wait(monitor, timeout);
				} catch (ThreadInterruptedException) {
					// .NET monitors can ignore notifications when there is a race between
					// the notification and the interrupt. So, we must regenerate the notification
					// if there are pending messages, in order to propagate the notification to
					// another blocked thread
					if (CanAcquire())
						Monitor.Pulse(monitor);
					throw;
				}
			} while (!CanReceive());
        	receivedMessage = ReceiveSideEffect();
			return true;
		}
    }

    // send a message to the queue
	public void Send(T sentMessage) {
        lock (monitor) {
        	UpdateStateOnSend(sentMessage);
			Monitor.Pulse(monitor);	/* only one thread cam receive the sent message */
		}
    }
}

File: Exams/Resoluções de Exames by Carlos Martins/generic-synch/GenericSynchronizerMonitorStyle.java

/**
 *  ISEL, LEIC, Concurrent Programming
 *
 *  Pseudo-code for implementing a generic synchronizer using the
 *  "Lampson & Redell monitor style" and Java.
 *
 *  Realizations for the case of semaphore and message queue synchronizers.
 *
 *  Note: To compile the java code contained in this file it will be necessary
 *        to comment the excerpts that contain pseudo-code.
 *
 *  Carlos Martins, October 2018
 */

import java.util.LinkedList;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.*;

/**
 * Generic classes used to express state, arguments, and results
 */

class SynchState {}
class InitializeArgs {}
class AcquireArgs {}
class ReleaseArgs {}
class AcquireResult {}

/**
 * This class implements a generic synchronizer using pseudo-code to express
 * data access and control synchronization (without considering the availability
 * of monitors).
 */

///***
class GenericSynchronizerMonitorStylePseudoCode {
	// lock that synchronizes the access to the mutable shared state
	private Lock lock = new Lock();
	
	// queue where block threads are (also protected by "lock") needed to
	// implement control synchronization inherent to acquire/release semantics.
	private WaitQueue waitQueue = new WaitQueue();
	
	// synchronization state
	private SynchState synchState;
	
	// initialize the synchronizer
	public GenericSynchronizerMonitorStylePseudoCode(InitialState initialState) {
		initialize "synchState" according to information specified by "initialState";
	}
	
	// test if synchronization state allows an immediate acquire
	private boolean canAcquire(AcquireArgs acquireArgs) {
		returns true if "syncState" satisfies an immediate acquire according to "acquireArgs";
	}
	
	// executes the processing associated with a successful acquire 
	private AcquireResult acquireSideEffect(AcquireArgs acquireArgs) {
		update "synchState" according to "acquireArgs" after a successful acquire;
		returns "the-proper-acquire-result"
	}
	
	// update synchronization state due to a release operation
	private void updateStateOnRelease(ReleaseArgs releaseArgs) {
		update "syncState" according to "releaseArgs";
	}
	
	public AcquireResult acquire(AcquireArgs acquireArgs) {
		lock.acquire();
		try {
			while (!canAcquire(acquireArgs)) {
				enqueue current thread on the "waitQueue" sensible to posterior wakeups;
				int depth = lock.releaseAll();
				block current thread until it is waked up by a releaser thread;
				lock.reAcquire(depth);
			}
			return acquireSideEffect(acquireArgs);
		} finally {
			lock.release();
		}
	}
	
	public void release(ReleaseArgs releaseArgs) {
		lock.acquire();
		try {
			updateStateOnRelease(releaseArgs);
			wakeup all the blocked threads that can have its acquire satisfied with the new value of "syncState";
		} finally {
			lock.Release();
		}
	}
}
//**/

/**
 * Generic synchronizer pseudo-code based on an *implicit Java monitor*, with
 * support for timeout on the acquire operation.
 */

///***
class GenericSynchronizerMonitorStyleImplicitMonitor {
	// implicit Java monitor provides synchronization of the access to the shared
	// mutable state, supports the control synchronization inherent to
	// acquire/release semantics.
	private final Object monitor = new Object();
    
	// synchronization state
	private SynchState synchState;

    // initialize the synchronizer
	public GenericSynchronizerMonitorStyleImplicitMonitor(InitializeArgs initialState) {
        initialize "synchState" according to information specified by "initialState";
    }

    // returns true if synchronization state allows an immediate acquire
    private boolean canAcquire(AcquireArgs acquireArgs) {
       returns true if "syncState" satisfies an immediate acquire according to "acquireArgs";
    }

    // executes the processing associated with a successful acquire 
    private AcquireResult acquireSideEffect(AcquireArgs acquireArgs) {
        update "synchState" according to "acquireArgs" after a successful acquire;
		returns "the-proper-acquire-result";
    }

    // update synchronization state due to a release operation
    private void updateStateOnRelease(ReleaseArgs releaseArgs) {
        // update "syncState" according to "releaseArgs";
    }

	// the generic acquire operation; returns null when it times out
    public AcquireResult acquire(AcquireArgs acquireArgs, long millisTimeout) throws InterruptedException {
        synchronized(monitor) {
			if (canAcquire(acquireArgs))
				return acquireSideEffect(acquireArgs);
			TimeoutHolder th = new TimeoutHolder(millisTimeout);
			do {
				if (th.isTimed()) {
					if ((millisTimeout = th.value()) <= 0)
						return null;		// timeout
					monitor.wait(millisTimeout);
				} else
					monitor.wait();
			} while (!canAcquire(acquireArgs));
			// successful acquire
        	return  acquireSideEffect(acquireArgs);
		}
    }

	// the generic release operation
    public void release(ReleaseArgs releaseArgs) {
		synchronized(monitor) {
        	updateStateOnRelease(releaseArgs);
			monitor.notifyAll();
			// or monitor.notify() if only one thread can have success in its acquire
		}
    }
}
//**/

/**
 * Generic synchronizer pseudo-code based on an *explicit Java monitor*, with
 * support for timeout on the acquire operation.
 */

///***
class GenericSynchronizerMonitorStyleExplicitMonitor {
	// explicit Java monitor that suports the synchronzation of shared data access
	// and supports also the control synchronization.
	private final Lock lock = new ReentrantLock();
	private final Condition okToAcquire = lock.newCondition(); 
	
	// synchronization state
	private SynchState synchState;

	// initialize the synchronizer
	public GenericSynchronizerMonitorStyleExplicitMonitor(InitializeArgs initialState) {
        initialize "synchState" according to information specified by "initialState";
    }

	// returns true if synchronization state allows an immediate acquire
	private boolean canAcquire(AcquireArgs acquireArgs) {
        returns true if "syncState" satisfies an immediate acquire according to "acquireArgs";
    }

	// executes the processing associated with a successful acquire
	private AcquireResult acquireSideEffect(AcquireArgs acquireArgs) {
        update "synchState" according to "acquireArgs" after a successful acquire;
		returns "the-proper-acquire-result";

    }

	// update synchronization state due to a release operation
	private void updateStateOnRelease(ReleaseArgs releaseArgs) {
        // update "syncState" according to "releaseArgs";
    }

	// generic acquire operation; returns null when it times out
	public AcquireResult acquire(AcquireArgs acquireArgs, long millisTimeout) throws InterruptedException {
		lock.lock();
		try {
			if (canAcquire(acquireArgs))
				return acquireSideEffect(acquireArgs);
			boolean isTimed = millisTimeout >= 0;
			long nanosTimeout = isTimed ? TimeUnit.MILLISECONDS.toNanos(millisTimeout) : 0L;
			do {
				if (isTimed) {
					if (nanosTimeout <= 0)
						return null; // timeout
					nanosTimeout = okToAcquire.awaitNanos(nanosTimeout);
				} else
					okToAcquire.await();
			} while (!canAcquire(acquireArgs));
			// successful acquire
			return acquireSideEffect(acquireArgs);
		} finally {
			lock.unlock();
		}
	}

	// generic release operation 
	public void release(ReleaseArgs releaseArgs) {
		lock.lock();
		try {
			updateStateOnRelease(releaseArgs);
			okToAcquire.signalAll();
			// or okToAcquire.signal() if only one thread can have success in its acquire
		} finally {
			lock.unlock();
		}
	}
}
//**/

/**
 * Semaphore following the monitor style, using an *implicit Java monitor*, with
 * support for timeout on the acquire operation.
 */

class SemaphoreMonitorStyleImplictMonitor {
	// implicit Java monitor that suports the synchronzation of shared data access
	// and supports also the control synchronization.
	private final Object monitor = new Object();
	
	// synchronization state
	private int permits;

	// initialize the semaphore
	public SemaphoreMonitorStyleImplictMonitor(int initial) {
		if (initial > 0)
			permits = initial;
	}

	// if the pending permits e equal or greater to the request,
	// we can acquire immediately
	private boolean canAcquire(int acquires) { return permits >= acquires; }

	// deduce the acquired permits
	private void acquireSideEffect(int acquires) { permits -= acquires; }

	// take into account the released permits
	private void updateStateOnRelease(int releases) { permits += releases; }

	// acquires the specified number of permits
	public boolean acquire(int acquires, long millisTimeout) throws InterruptedException {
		synchronized(monitor) {
			if (canAcquire(acquires)) {
				acquireSideEffect(acquires);
				return true;
			}
			TimeoutHolder th = new TimeoutHolder(millisTimeout);
			do {
				if (th.isTimed()) {
					if ((millisTimeout = th.value()) <= 0)
						return false; // timeout
					monitor.wait(millisTimeout);
				} else
					monitor.wait();
			} while (!canAcquire(acquires));
			// successful acquire
			acquireSideEffect(acquires);
			return true;
		}
	}

	// releases the specified number of permits
	public void release(int releases) {
		synchronized(monitor) {
			updateStateOnRelease(releases);
			monitor.notifyAll();
		}
	}
}

/**
 * Semaphore following the monitor style, using an *explicit Java monitors*,
 * with timeout support on acquire operation.
 */

 class SemaphoreMonitorStyleExplicitMonitor {
	// explicit Java monitor that suports the synchronzation of shared data access
	// and supports also the control synchronization.
	private final Lock lock = new ReentrantLock();
	private final Condition okToAcquire = lock.newCondition();

	// synchronization state
	private int permits;
	
	// initialize the semaphore
	public SemaphoreMonitorStyleExplicitMonitor(int initial) {
		if (initial > 0)
			permits = initial;
	}
	
	// if the pending permits e equal or greater to the request,
	// we can acquire immediately
	private boolean canAcquire(int acquires) { return permits >= acquires; }
	
	// deduce the acquired permits
	private void acquireSideEffect(int acquires) { permits -= acquires; }
	
	// take into account the released permits
	private void updateStateOnRelease(int releases) { permits += releases; }
	
	// acquires the specified number of permits
	public boolean acquire(int acquires, long millisTimeout) throws InterruptedException {
		lock.lock();
		try {
			if (canAcquire(acquires)) {
                acquireSideEffect(acquires);
				return true;
			}
			boolean isTimed = millisTimeout > 0;
			long nanosTimeout = isTimed ? TimeUnit.MILLISECONDS.toNanos(millisTimeout) : 0L;
			do {
				if (isTimed) {
					if (nanosTimeout <= 0)
						return false;
					nanosTimeout = okToAcquire.awaitNanos(nanosTimeout);
				} else
					okToAcquire.await();
			} while (!canAcquire(acquires));
			acquireSideEffect(acquires);
			return true;
		} finally {
			lock.unlock();
		}
	}

	// releases the specified number of permits
	public void release(int releases) {
		lock.lock();
		try {
			updateStateOnRelease(releases);
			okToAcquire.signalAll();	// this release can satisfy multiple acquires
		} finally {
			lock.unlock();
		}
	}
}

/**
 * Message queue following the monitor style, using a *Java implicit monitors*,
 * with timeout support on the receive acquire operation.
 */
 
class MessageQueueMonitorStyleImplicitMonitor<T> {
 	// implicit Java monitor that suports the synchronzation of shared data access
	// and supports also the control synchronization.
	private final Object monitor = new Object();
	
	// list of messages pending for reception
 	private final LinkedList<T> pendingMessages = new LinkedList<T>();

    // initializes the message queue
	public MessageQueueMonitorStyleImplicitMonitor() {}

    // if there is an pending message, receive cam succeed immediately
    private boolean canReceive() { return pendingMessages.size() > 0; }

    // when a message is received, it must be removed from the pending message list
	private T receiveSideEffect() { return pendingMessages.poll(); }

    // add the sent message to the pending messages list
    private void updateStateOnSend(T sentMessage) { pendingMessages.addLast(sentMessage); }

    // receive the next message; returns null when it times out
	public T receive(long millisTimeout) throws InterruptedException {
        synchronized(monitor) {
			if (canReceive())
				return receiveSideEffect();
			TimeoutHolder th = new TimeoutHolder(millisTimeout);
			do {
				if (th.isTimed()) {
					if ((millisTimeout = th.value()) <= 0)
						return null;	// timeout
					monitor.wait(millisTimeout);
				} else
					monitor.wait(millisTimeout);
			} while (!canReceive());
        	return receiveSideEffect();
		}
    }

    // send a message to the queue
	public void send(T sentMessage) {
        synchronized(monitor) {
        	updateStateOnSend(sentMessage);
			monitor.notify();	/* only one thread cam receive the sent message */
		}
    }
}

/**
 * Message queue following the monitor style, using an *explicit Java monitor*,
 * with timeout support on the receive operation.
 */

class MessageQueueMonitorStyleExplicitMonitor<T> {
	// explicit Java monitor that suports the synchronzation of shared data access
	// and supports also the control synchronization.
	private final Lock lock = new ReentrantLock();
	private final Condition okToReceive = lock.newCondition();

	// synchronization state: list of messages pending for reception
	private LinkedList<T> pendingMessages = new LinkedList<T>();

	// initializes the message queue
	public MessageQueueMonitorStyleExplicitMonitor() {}

	// if there is an pending message, receive cam succeed immediately
	private boolean canReceive() { return pendingMessages.size() > 0; }

	// when a message is received, it must be removed from the pending message list
	private T receiveSideEffect() { return pendingMessages.poll(); }

	// add the sent message to the pending messages list
	private void updateStateOnSend(T sentMessage) { pendingMessages.addLast(sentMessage); }

	// receive the next message; returns null when it times out
	public T receive(long millisTimeout) throws InterruptedException {
		lock.lock();
		try {
			if (canReceive())
				return receiveSideEffect();
			boolean isTimed = millisTimeout > 0;
			long nanosTimeout = isTimed ? TimeUnit.MILLISECONDS.toNanos(millisTimeout) : 0L;
			do {
				if (isTimed) {
					if (nanosTimeout <= 0)
						return null; // timeout
					nanosTimeout = okToReceive.awaitNanos(nanosTimeout);
				} else
					okToReceive.await();
			} while (!canReceive());
			return receiveSideEffect();
		} finally {
			lock.unlock();
		}
	}
	
	// send a message to the queue
	public void send(T sentMessage) {
		lock.lock();
		try {
			updateStateOnSend(sentMessage);
			okToReceive.signal(); 	/* only one thread cam receive the sent message */
		} finally {
			lock.unlock();
		}
	}
}
